PHPCS rule to prefer implode over join

markbiek opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
<!-- Would you like this feature to be tested by Beta testers?
Please add testing instructions to projects/plugins/jetpack/to-test.md in a new commit as part of your PR. -->
<!-- a12s: If you have an expected version that you're aiming for the PR to add, please use the Milestone field to communicate it. If you leave it blank, that indicates there isn't a preference. -->

## Proposed changes:
<!--- Explain what functional changes your PR includes -->
This adds a phpcs rule to enforce using `implode` instead of `join`.

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion
<!-- If you're an Automattician, include a shortlink to the p2 discussion with Jetpack Product here. -->
<!-- Make sure any changes to existing products have been discussed and agreed upon -->

## Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->

## Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

* Check out this branch.
* Make an update that uses the `join` function and try to commit it.
* Your code should automatically be updated to use `implode` instead of `join`.

anomiex commented:
Seems to work, as can be seen by all the failures in the linting job. 😀

Two suggestions:

* You linked elsewhere to [a list of other such replacements](https://github.com/Automattic/wpcom-php-migration/blob/main/rector.php#L57) that wpcom is wanting to apply. We should probably do all of those.
* This rule does not fix with phpcbf. Perhaps we should use [MediaWiki.Usage.ForbiddenFunctions](https://github.com/wikimedia/mediawiki-tools-codesniffer/blob/v40.0.0/MediaWiki/Sniffs/Usage/ForbiddenFunctionsSniff.php) instead, which does have a fixer? We already use other MediaWiki phpcs rules so it's already available and just needs activating.
  * It looks like that one does have more replacements built in though. Looking through the code, it looks like the way to disable the built-in fixes (once we update to 41.0.0, which I've just told Renovate to make a PR for) is to set the property to map the function to itself.

markbiek commented:
@anomiex Could you give me a little more context on this? I'm not sure I follow what you're describing.

> Looking through the code, it looks like the way to disable the built-in fixes (once we update to 41.0.0, which I've just told Renovate to make a PR for) is to set the property to map the function to itself.

I _think_ this is doing what you're asking for but I want to make sure I understand
```
<rule ref="MediaWiki.ForbiddenFunctions">
        <properties>
            <property name="forbidden_functions">
                join=implode
            </property>
        </properties>
    </rule>
```

Assuming that's what you're talking about, I can add the other functions that rector is checking for.

markbiek committed:
[not verified] Update list of forbidden functions

These are all of the function replacements we're doing with rector as
part of the PHP8 compatibility work.

markbiek committed:
changelog

nelsonec87 commented on projects/packages/codesniffer/Jetpack/ruleset.xml:
```diff
@@ -84,4 +84,32 @@
+        <properties>
+            <property name="forbidden_functions">
+                join=implode
+				chop=rtrim
```
indentation?

nelsonec87 reviewed (dismissed):
LGTM!

markbiek committed:
Fix indentation

nelsonec87 reviewed (dismissed):
LGTM

markbiek committed:
Version bump

markbiek committed:
Update phpunit configuration schema

markbiek commented:
@anomiex I'm not sure what do do about that "PHP 8.2 WP latest" test failure. Do you have any advice?

anomiex committed:
Fix rule configuration

anomiex committed:
remove unintended change

anomiex committed:
Add tests that new rule is working

anomiex committed:
Update root composer.lock

anomiex commented:
The PHP 8.2 failure was because you had the configuration wrong. I fixed it in 7626f8f30014871b512ca9bf3fe98b1d84089b47.

anomiex committed:
Auto-fix the auto-fixable

chop→rtrim fputs→fwrite is_writeable→is_writable join→implode sizeof→count

anomiex committed:
Fix other sniffs that complain on the fixed lines

anomiex committed:
Replace multi-arg `isset()` for better clarity and new phpcs sniff

anomiex committed:
Merge remote-tracking branch 'origin/trunk' into update/phpcs-rules

anomiex committed:
Fix two more WordPress.WP.AlternativeFunctions

anomiex commented:
Looks good to me now, but someone other than me should review all the fixes I made.

@markbiek If you're the one who does that review, you can't hit "approve" in GitHub because you're listed as the author of the PR. But if you comment to say my changes are good, then I can hit the button.

markbiek commented:
> Looks good to me now, but someone other than me should review all the fixes I made.
> 
> @markbiek If you're the one who does that review, you can't hit "approve" in GitHub because you're listed as the author of the PR. But if you comment to say my changes are good, then I can hit the button.

Everything looks good to me!

anomiex commented:
For the (internal) record, more details have been posted at pdWQjU-j5-p2.
